### PR TITLE
[Brig] Move password verification to the AuthenticationSubsystem, move to Argon2id with new settings.

### PR DESCRIPTION
**Pull request by elland**
https://wearezeta.atlassian.net/browse/WPB-9746

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/brig/brig.integration.yaml [hunk 1]
```diff
@@ -171,7 +171,7 @@ optSettings:
     timeout: 5  # seconds.  if you reach the limit, how long do you have to wait to try again.
     retryLimit: 5  # how many times can you have a failed login in that timeframe.
   setSuspendInactiveUsers:  # if this is omitted: never suspend inactive users.
-    suspendTimeout: 4
+    suspendTimeout: 10
```
Comment (mdimjasevic) [inline services/brig/brig.integration.yaml:174]: Why this change?

File: services/brig/src/Brig/API/OAuth.hs [hunk 2]
```diff
@@ -100,6 +106,7 @@ registerOAuthClient (OAuthClientConfig name uri) = do
     createSecret :: (MonadIO m) => m OAuthClientPlainTextSecret
     createSecret = OAuthClientPlainTextSecret <$> rand32Bytes
 
+    -- TODO(elland): figure out why
```
Comment (mdimjasevic) [inline services/brig/src/Brig/API/OAuth.hs:109]: Something still to do.

**Review by mdimjasevic — COMMENTED** [review]
Here is a partial review

**Comment by mdimjasevic** [comment]
Have you searched for all usages of `mkSafePasswordScrypt` and checked that's as expected?

File: services/brig/src/Brig/Data/User.hs [hunk 3]
```diff
@@ -180,17 +181,24 @@ newAccountInviteViaScim uid externalId tid locale name email = do
...
 -- | Mandatory password authentication.
-authenticate :: forall r. (Member PasswordStore r) => UserId -> PlainTextPassword6 -> ExceptT AuthError (AppT r) ()
+authenticate ::
+  forall r.
+  (Member PasswordStore r, Member AuthenticationSubsystem r) =>
+  UserId ->
+  PlainTextPassword6 ->
+  ExceptT AuthError (AppT r) ()
 authenticate u pw =
+  -- TODO: Move this logic into auth subsystem.
```
Comment (mdimjasevic) [inline services/brig/src/Brig/Data/User.hs:191]: Still something to do.

File: services/brig/src/Brig/Data/User.hs [hunk 4]
```diff
@@ -225,10 +231,10 @@ reauthenticate u pw =
   where
     maybeReAuth pw' = case pw of
       Nothing -> do
-        musr <- lookupUser NoPendingInvitations u
+        musr <- wrapClientE $ lookupUser NoPendingInvitations u
```
Comment (mdimjasevic) [inline services/brig/src/Brig/Data/User.hs:234]: Don't we have an equivalent in the `UserStore` effect?

File: services/brig/src/Brig/Provider/API.hs [hunk 5]
```diff
@@ -265,20 +273,31 @@ activateAccountKey key val = do
       lift $ sendApprovalConfirmMail name email
       pure . Just $ Public.ProviderActivationResponse email
 
-getActivationCodeH :: (Member GalleyAPIAccess r, Member VerificationCodeSubsystem r) => EmailAddress -> (Handler r) Code.KeyValuePair
+getActivationCodeH ::
```
Comment (mdimjasevic) [inline services/brig/src/Brig/Provider/API.hs:276]: We could drop this `wai-routes` convention of ending handler names with "H" now that you've already changed the line.

**Review by mdimjasevic — COMMENTED** [review]
Some more comments inlined.

Given that this is changing the default hashing algorithm, and we forgot to do it before in some places, what would be good tests to add to capture this change?

File: libs/wire-subsystems/src/Wire/PasswordStore/Cassandra.hs [hunk 6]
```diff
@@ -16,6 +16,12 @@ interpretPasswordStore casClient =
     runEmbedded (runClient casClient) . \case
       UpsertHashedPassword uid password -> embed $ updatePasswordImpl uid password
       LookupHashedPassword uid -> embed $ lookupPasswordImpl uid
+      LookupHashedProviderPassword pid -> embed $ lookupProviderPasswordImpl pid
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/wire-subsystems/src/Wire/PasswordStore/Cassandra.hs:19]: Yes, different table.

File: libs/wire-api/src/Wire/API/Password.hs [hunk 7]
```diff
@@ -352,3 +355,12 @@ instance ToSchema PasswordReqBody where
...
         <$> fromPasswordReqBody .= maybe_ (optField "password" schema)
+
+-------------------------------------------------------------------------------
+-- Generate test passwords, benchmark
+
+genTestPasswords :: IO [(Text, Text)]
+genTestPasswords = replicateM 100 do
+  pwd <- genPassword
+  hash <- mkSafePassword pwd
+  pure (fromPlainTextPassword pwd, fromPassword hash)
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/wire-api/src/Wire/API/Password.hs:366]: Requires access to private functions, unsafe to deconstruct the opaque password type. I should document it.

File: libs/wire-api/src/Wire/API/Password.hs [hunk 8]
```diff
@@ -168,7 +171,7 @@ hashPasswordScrypt password = do
 
 hashPasswordArgon2id :: (MonadIO m) => ByteString -> m Text
 hashPasswordArgon2id pwd = do
-  salt <- newSalt 32
+  salt <- newSalt 16
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/wire-api/src/Wire/API/Password.hs]: It isn't super clear but it's in the ticket description, which params we want to use :)

File: libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs [hunk 9]
```diff
@@ -23,14 +23,18 @@ import Data.Misc
...
 import Polysemy
+import Wire.API.Password (Password, PasswordStatus)
 import Wire.API.User
-import Wire.API.User.Password
+import Wire.API.User.Password (PasswordResetCode, PasswordResetIdentity)
 import Wire.UserKeyStore
 
 data AuthenticationSubsystem m a where
-  VerifyPassword :: Local UserId -> PlainTextPassword6 -> AuthenticationSubsystem m ()
+  VerifyPasswordE :: Local UserId -> PlainTextPassword6 -> AuthenticationSubsystem m ()
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs:32]: `Error`, I'll make it more explicit.

File: libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs [hunk 10]
```diff
@@ -23,14 +23,18 @@ import Data.Misc
...
   CreatePasswordResetCode :: EmailKey -> AuthenticationSubsystem m ()
   ResetPassword :: PasswordResetIdentity -> PasswordResetCode -> PlainTextPassword8 -> AuthenticationSubsystem m ()
+  VerifyPassword :: PlainTextPassword6 -> Password -> AuthenticationSubsystem m (Bool, PasswordStatus)
+  VerifyUserPassword :: UserId -> PlainTextPassword6 -> AuthenticationSubsystem r (Bool, PasswordStatus)
+  VerifyProviderPassword :: ProviderId -> PlainTextPassword6 -> AuthenticationSubsystem r (Bool, PasswordStatus)
```
[the remark this answers is not in the document]
Comment (elland) [reply libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs:37]: It isn't, but since we never forced users or providers to migrate to pwd8, we just prompt that lazily as they re-auth, we need to keep supporting it.

File: Makefile [hunk 11]
```diff
@@ -62,6 +62,10 @@ full-clean: clean
 	clean-rabbit
 	@echo -e "\n\n*** NOTE: you may want to also 'rm -rf ~/.cabal/store \$$CABAL_DIR/store', not sure.\n"
 
+.PHONY: clean-rabbit
+clean-rabbit:
+	rabbitmqadmin -f pretty_json list queues vhost name messages | jq -r '.[] | "rabbitmqadmin delete queue name=\(.name) --vhost=\(.vhost)"' | bash
+
```
Comment (akshaymankar) [inline Makefile]: This already exists few lines above

```suggestion
```

File: libs/wire-api/src/Wire/API/Password.hs [hunk 12]
```diff
@@ -1,4 +1,6 @@
 {-# LANGUAGE RecordWildCards #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
```
Comment (akshaymankar) [inline libs/wire-api/src/Wire/API/Password.hs:3]: We shouldn't use this in prod code.

```suggestion
```

File: libs/wire-api/src/Wire/API/Password.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply libs/wire-api/src/Wire/API/Password.hs:174]: We should perhaps write down all these defaults in one place with an explanation. There is a comment from @sanojwr on the ticket which says we should use use 128 bit salt. The RFC also suggests using 16 bytes: https://www.rfc-editor.org/rfc/rfc9106.html#section-3.1-1

File: libs/wire-api/src/Wire/API/Password.hs [hunk 13]
```diff
@@ -277,12 +280,12 @@ parseScryptPasswordHashParams passwordHash = do
 
 hashPasswordWithOptions :: Argon2idOptions -> ByteString -> ByteString -> ByteString
 hashPasswordWithOptions opts password salt =
-  case (Argon2.hash opts password salt 64) of
-    -- CryptoFailed occurs when salt, output or input are too small/big.
-    -- since we control those values ourselves, it should never have a runtime error
-    -- unless we've caused it ourselves.
-    CryptoFailed cErr -> error $ "Impossible error: " <> show cErr
-    CryptoPassed hash -> hash
+  let tagSize = 16
```
Comment (akshaymankar) [inline libs/wire-api/src/Wire/API/Password.hs:283]: The comment from @sanojwr mentions 256 bit tag size, why 16?

File: libs/wire-api/src/Wire/API/Password.hs (hunk 7 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply libs/wire-api/src/Wire/API/Password.hs:366]: This function is unused actually.

File: libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs [hunk 14]
```diff
@@ -23,14 +23,18 @@ import Data.Misc
...
 import Polysemy
+import Wire.API.Password (Password, PasswordStatus)
 import Wire.API.User
-import Wire.API.User.Password
+import Wire.API.User.Password (PasswordResetCode, PasswordResetIdentity)
 import Wire.UserKeyStore
 
 data AuthenticationSubsystem m a where
-  VerifyPassword :: Local UserId -> PlainTextPassword6 -> AuthenticationSubsystem m ()
+  VerifyPasswordError :: Local UserId -> PlainTextPassword6 -> AuthenticationSubsystem m ()
```
Comment (akshaymankar) [inline libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs:32]:
```suggestion
  VerifyUserPasswordError :: Local UserId -> PlainTextPassword6 -> AuthenticationSubsystem m ()
```

Given there is `VerifyPassword` and `VerifyUserPassword`, I think its less confusing to call this `VerifyUserPassword`.

Also would look nicer if these things were written next to each other.

File: libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs [hunk 15]
```diff
@@ -23,14 +23,18 @@ import Data.Misc
...
   CreatePasswordResetCode :: EmailKey -> AuthenticationSubsystem m ()
   ResetPassword :: PasswordResetIdentity -> PasswordResetCode -> PlainTextPassword8 -> AuthenticationSubsystem m ()
+  VerifyPassword :: PlainTextPassword6 -> Password -> AuthenticationSubsystem m (Bool, PasswordStatus)
```
Comment (akshaymankar) [inline libs/wire-subsystems/src/Wire/AuthenticationSubsystem.hs:35]: From perspective of the AuthenticationSubsystem API this looks weird, why does anyone else have access to the hashed password outside this subsystem?

File: libs/wire-subsystems/src/Wire/AuthenticationSubsystem/Interpreter.hs [hunk 16]
```diff
@@ -266,3 +262,44 @@ resetPasswordImpl ident code pw = do
...
+verifyProviderPasswordImpl ::
+  (Member PasswordStore r, Member (Error AuthenticationSubsystemError) r) =>
+  ProviderId ->
+  PlainTextPassword6 ->
+  Sem r (Bool, PasswordStatus)
+verifyProviderPasswordImpl pid plaintext = do
+  -- We type-erase uid here
+  password <-
+    PasswordStore.lookupHashedProviderPassword pid
+      >>= maybe (throw AuthenticationSubsystemBadCredentials) pure
```
[the remark this answers is not in the document]
Comment (mdimjasevic) [reply libs/wire-subsystems/src/Wire/AuthenticationSubsystem/Interpreter.hs:282]: The call chain ain't long here. Note that this is in the interpreter, not the application code. I expect we call this interpreter only once.

File: libs/wire-api/src/Wire/API/Password.hs [hunk 17]
```diff
@@ -191,96 +218,100 @@ hashPasswordArgon2idWithOptions opts salt pwd = do
...
+    Left argon2ParseError ->
       case parseScryptPasswordHashParams $ Text.encodeUtf8 expected of
-        Just (sparams, saltS, hashedKeyS) ->
-          let producedKeyS = hashPasswordWithParams sparams (Text.encodeUtf8 actual) saltS
-           in (hashedKeyS `constEq` producedKeyS, PasswordStatusNeedsUpdate)
-        Nothing -> (False, PasswordStatusNeedsUpdate)
+        Right hashedPassword -> Right $ ScryptPassword hashedPassword
+        Left scryptParseError ->
+          Left $
+            "failed to parse password are argon2 or scrypt hashed password: argon2 parse error: "
```
Comment (battermann) [inline libs/wire-api/src/Wire/API/Password.hs]:
```suggestion
            "failed to parse password argon2 or scrypt hashed password: argon2 parse error: "
```

**Review by battermann — APPROVED** [review]

File: changelog.d/5-internal/pwd [hunk 18]
```diff
@@ -0,0 +1 @@
+Changed default password hashing from Scrypt to Argon2id.
```
[the remark this answers is not in the document]
Comment (fisx) [reply changelog.d/5-internal/pwd:1]: (also, if you do an extensive refactoring, it should be mentioned in the changelog.)
